### PR TITLE
feat: Add manual stop/start for EC2 and SageMaker SC workspaces 

### PR DESCRIPTION
**Pull request by Bingjiling**
Issue #, if available:

Description of changes:
* Stop and start workflow for EC2 and SageMaker SC workspaces 
* Update EC2 IP and DNS data in DDB on EC2 manual start 
* Deployed to an AWS account and tested with Postman 
* UI changes

Checklist: 

* [x] Have you successfully deployed to an AWS account with your changes? 
* [x] Have you linted your code locally prior to submission?
* [x] Have you successfully ran unit tests and manual tests with your changes locally?
(../main/end-to-end-tests/README.md)?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js [hunk 1]
```diff
@@ -271,6 +275,89 @@ class EnvironmentScService extends Service {
...
+    const expectedStatus = operation === 'start' ? 'STOPPED' : 'COMPLETED';
+    if (status !== expectedStatus) {
+      throw this.boom.badRequest(
+        `unable to ${operation} environment with id "${id}" - current status "${status}"`,
+        true,
+      );
+    }
+    let instanceType;
+    let instanceIdentifier;
+    const outputsObject = cfnOutputsToObject(outputs);
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:300]: outputsObject is a dictionary. Let's name it something like 'outputsDict'

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js [hunk 2]
```diff
@@ -101,14 +101,6 @@ class StartSagemakerEnvironment extends StepBase {
 
   async getSageMakerService() {
     const [aws] = await this.mustFindServices(['aws']);
-    // increase retry defaults to reduce the risk of failures caused
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js:104]: Why was this part removed?

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+id: st-stop-ec2-environment-sc
+v: 1
+title: Stop EC2 Environment
```
Comment (SanketD92) [inline addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.yml:3]: Optional: Add the phrase 'Service Catalog' in title and desc in these new files.

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js (hunk 2 above)
Comment (Bingjiling) [reply addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js:104]: Ah! This was a mix up, I was trying to remove this part from start-sagemaker-enviornment-sc.js. Good catch!

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js [hunk 4]
```diff
@@ -101,14 +101,6 @@ class StartSagemakerEnvironment extends StepBase {
...
-    // by throttles
-    aws.sdk.config.update({
-      maxRetries: 6, // default 3
-      retryDelayOptions: {
-        base: 300, // default 100
-      },
-    });
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js:111]: Why was this part removed?

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js (hunk 1 above)
Comment (Bingjiling) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:300]: Emmmm there is no 'dictionary' type in JS, a JS object is essentially key value pairs that can act as a dictionary. I'll update the function name to be something like cfnOutputsArrayToObject()?

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js [hunk 5]
```diff
@@ -271,6 +275,89 @@ class EnvironmentScService extends Service {
...
 
+  async changeWorkspaceRunState(requestContext, { id, operation }) {
+    const existingEnvironment = await this.mustFind(requestContext, { id });
+
+    // Make sure the user has permissions to change the environment run state
+    await this.assertAuthorized(
+      requestContext,
+      { action: 'update-sc', conditions: [this._allowAuthorized] },
+      existingEnvironment,
+    );
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:286]: Is the user which is being granted authorization same as the user which provisioned the EC2 instance in the first place?

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    const ec2 = await this.getEc2Service();
+
+    const params = {
+      InstanceIds: [instanceId],
+    };
+
+    let instanceStatusInfo;
+    try {
+      const paramsForDescribe = { ...params, IncludeAllInstances: true };
+      instanceStatusInfo = await ec2.describeInstanceStatus(paramsForDescribe).promise();
```
Comment (jn1119) [inline addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:38]: What is the behavior of describe API if the instanceId doesn't exist?

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('describe ec2 instance error: ', error);
+      throw error;
+    }
+
+    const status = _.get(instanceStatusInfo, 'InstanceStatuses[0].InstanceState.Name').toUpperCase();
+
+    if (!['STOPPING', 'STOPPED'].includes(status)) {
```
Comment (jn1119) [inline addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:47]: If the status is 'STOPPED' we can simply return, no?

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+const _ = require('lodash');
+const StepBase = require('@aws-ee/base-workflow-core/lib/workflow/helpers/step-base');
+
+class StartEc2EnvironmentSc extends StepBase {
```
Comment (jn1119) [inline addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:19]: Can we add some unit tests for all the new steps?

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js (hunk 4 above)
Comment (Bingjiling) [reply addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/start-sagemaker-environment/start-sagemaker-environment.js:111]: Ah! This was a mix up. I was trying to remove this part from start-sagemaker-enviornment-sc.js. Will update. Good catch!

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js [hunk 9]
```diff
@@ -271,6 +275,89 @@ class EnvironmentScService extends Service {
...
+    const workflowTriggerService = await this.service('workflowTriggerService');
+    const input = {
+      environmentId: existingEnvironment.id,
+      instanceIdentifier,
+      requestContext,
+      cfnExecutionRole,
+      roleExternalId,
+    };
+
+    await workflowTriggerService.triggerWorkflow(requestContext, meta, input);
```
[the remark this answers is not in the document]
Comment (Bingjiling) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:342]: Yes, this triggers the workflow with the name specified in 'meta', will add comment for clarification.

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js (hunk 8 above)
Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:19]: Yes! Will do.

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      instanceStatusInfo = await ec2.describeInstanceStatus(paramsForDescribe).promise();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('describe ec2 instance error: ', error);
+      throw error;
+    }
+
+    const status = _.get(instanceStatusInfo, 'InstanceStatuses[0].InstanceState.Name').toUpperCase();
+
+    if (status !== 'STOPPED') {
```
[the remark this answers is not in the document]
Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:47]: The API is idempotent now since the environment is "locked" in STARTING status in DDB table before the workflow is invoked. Other requests coming in won't be able to trigger the workflow since they will check DDB first, see it in STARTING status, and errors out. The only possibility here is the instance was started outside of SWB and it is out of sync with DDB table. I'm not sure if the logic should try to update the DDB table status to RUNNING if that's the case? (Apologize if it does not make a whole lot of sense. We can do a quick call if needed : )

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+      // eslint-disable-next-line no-console
+      console.error('start ec2 instance error: ', error);
+      throw error;
+    }
+
+    await this.updateEnvironment({ status: 'STARTING' });
+    this.state.setKey('STATE_INSTANCE_ID', instanceId);
+
+    return this.wait(5)
+      .maxAttempts(120)
```
[the remark this answers is not in the document]
Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:63]: That's a good question! I'll look into how maxAttempts and wait works.

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js (hunk 5 above)
Comment (Bingjiling) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:286]: Both the user that provisioned the instance and Admin has access to update the environment. The implementation is here -> https://github.com/awslabs/service-workbench-on-aws/blob/master/addons/addon-base-raas/packages/base-raas-services/lib/environment/environment-authz-service.js#L54

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js (hunk 6 above)
Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:38]: It would throw and invalid-id exception when the instanceId does not exist.

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (hatimkhan) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:63]: Most of the time, the AWS SDK will handle the backoff logic for you.  For example, certain AWS APIs will return errors that are marked as 'retryable'.  When the AWS SDK sees that in the exception, it will retry again with exponential backoff logic.  If you still want to introduce some randomness into the wait.  You can take a look at  addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/delete-environment/delete-environment.js (line 76), using the fuzz utility function.

`return this.wait(fuzz(80))
      .maxAttempts(120)`

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jn1119) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js:47]: Synced up over a call. I didn't realize we were also updating the DynamoDB status here. My main comment was around not executing `startInstance` and `checkInstanceStarted` methods since the instance was already in `start ` state. We can update DynamoDB directly.

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    const environment = {
+      id,
+      rev: existingEnvRecord.rev || 0,
+      status,
+    };
+    await environmentScService.update(requestContext, environment);
+  }
+
+  async onFail() {
+    return this.updateEnvironmentStatus('STOPPING_FAILED');
```
Comment (jn1119) [inline addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:145]: It might be out of scope but can user retry start/stop functionality if they get into this failed state due to network or EC2 issues?

Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:145]: Good call! I'll see if I can slip it in this PR :)

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js (hunk 1 above)
Comment (SanketD92) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:300]: Since this would be clear to people working with JS, I wouldn't mind this being left as-is. Thanks for clarifying

File: addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js (hunk 12 above)
Comment (Bingjiling) [reply addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js:145]: Looks like I don't have to add this one in. I'll create a separate ticket for it.

File: addons/addon-base-raas-ui/packages/base-raas-ui/src/models/environments-sc/ScEnvironmentsStore.js [hunk 13]
```diff
@@ -130,6 +157,22 @@ const ScEnvironmentsStore = BaseStore.named('ScEnvironmentsStore')
...
+      const outputs = self.environments.get(id).outputs;
+      let result = false;
+      // TODO: Remove isRStudio check after CNAME patch
+      let isRStudio = false;
+      outputs.forEach(output => {
+        if (output.OutputKey === 'Ec2WorkspaceInstanceId' || output.OutputKey === 'NotebookInstanceName') {
+          result = true;
+        }
+        if (output.OutputKey === 'MetaConnection1Type') {
+          isRStudio = true;
```
Comment (SanketD92) [inline addons/addon-base-raas-ui/packages/base-raas-ui/src/models/environments-sc/ScEnvironmentsStore.js:170]: It's not necessary that every workspace with a MetaConnection1Type output is an RStudio. 

Here you can only determine if this is an RStudio environment if an output with Key 'MetaConnection1Type' exists AND its Value is 'RStudio'

File: addons/addon-base-raas-ui/packages/base-raas-ui/src/models/environments-sc/ScEnvironmentsStore.js [hunk 14]
```diff
@@ -130,6 +157,22 @@ const ScEnvironmentsStore = BaseStore.named('ScEnvironmentsStore')
       return self.environments.get(id);
     },
 
+    canChangeState(id) {
+      const outputs = self.environments.get(id).outputs;
+      let result = false;
+      // TODO: Remove isRStudio check after CNAME patch
+      let isRStudio = false;
+      outputs.forEach(output => {
+        if (output.OutputKey === 'Ec2WorkspaceInstanceId' || output.OutputKey === 'NotebookInstanceName') {
```
Comment (SanketD92) [inline addons/addon-base-raas-ui/packages/base-raas-ui/src/models/environments-sc/ScEnvironmentsStore.js:166]: Since you're explicitly checking for RStudio in the next if block, I think you can safely remove the check for 'Ec2WorkspaceInstanceId' here

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js [hunk 15]
```diff
@@ -271,6 +275,103 @@ class EnvironmentScService extends Service {
...
+      throw this.boom.badRequest(
+        `unable to ${operation} environment with id "${id}" - current status "${status}"`,
+        true,
+      );
+    }
+    let instanceType;
+    let instanceIdentifier;
+    const outputsObject = cfnOutputsArrayToObject(outputs);
+    // TODO: Remove MetaConnection1Type check to include RStudio after CNAME patch
+    if ('Ec2WorkspaceInstanceId' in outputsObject && !('MetaConnection1Type' in outputsObject)) {
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js:313]: Same comment applies here. Having the 'MetaConnection1Type' output available might not necessarily mean it's an RStudio environment. The value for this needs to be 'RStudio'. 
(Users can add templates with this output for a non-RStudio workspace type.)

**Review by SanketD92 — APPROVED** [review]
LGTM
